### PR TITLE
Add PME slab correction support

### PR DESCRIPTION
**Pull request by atulcthakur**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Adds 2D slab correction support to the PME path, matching the Ewald slab behavior and standalone slab correction API.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Changes Made

- Added slab_correction support to `particle_mesh_ewald`.
- Added PME slab tests for energies, forces, charge gradients, virials, dtype preservation, hybrid charge gradients, and TorchPME cross-validation.
- Updated slab docs, examples, and electrostatics benchmark options for Ewald/PME slab usage.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

**Comment by atulcthakur** [comment]
/ok to test https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/19108c962f0409874e100a1c105796edcc3f497e

**Comment by atulcthakur** [comment]
/ok to test [7262223](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/7262223ef6955eeeca9094ca407ed728d39dd7e0)

**Comment by atulcthakur** [comment]
/ok to test [0446105](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/0446105c0443e1bc24e2d16e95ef2d032350cc59)

**Comment by atulcthakur** [comment]
/ok to test https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/54a967ded1e5cefcf5073abadb776014b5bac2bc

File: nvalchemiops/torch/interactions/electrostatics/pme.py [hunk 1]
```diff
@@ -2319,21 +2345,97 @@ def particle_mesh_ewald(
         hybrid_forces=hybrid_forces,
     )
 
-    # Normalize return tuples for easy combination
-    # Both rs and rec return: energies, [forces], [charge_grads], [virial]
-    # where virial is always last if present
+    slab_tuple: tuple[torch.Tensor, ...] | None = None
+    if slab_correction:
```
Comment (laserkelvin) [inline nvalchemiops/torch/interactions/electrostatics/pme.py:2363]: Can we validate `pbc` before launching the real/reciprocal PME work? With `slab_correction=True` and missing/invalid `pbc`, we currently do the real-space and FFT-heavy reciprocal path before `apply_slab_correction` raises. A cheap fail-fast check right after `_prepare_cell()` would save expensive bad-argument runs; ideally reuse the same slab `pbc` validation used by `apply_slab_correction`.

**Review by laserkelvin — COMMENTED** [review]
Left a few focused comments on PME slab API/perf consistency and test coverage. Overall this looks consistent with the Ewald slab API from PR #78.

File: nvalchemiops/torch/interactions/electrostatics/pme.py [hunk 2]
```diff
@@ -2319,21 +2345,97 @@ def particle_mesh_ewald(
...
-        results.append(r + s)
+    if compute_forces:
+        forces = rs_tuple[tuple_index] + rec_tuple[tuple_index]
+        if slab_tuple is not None:
+            forces = forces + slab_tuple[tuple_index]
+        results += (forces,)
+        tuple_index += 1
+
+    if compute_charge_gradients:
+        charge_grads = rs_tuple[tuple_index] + rec_tuple[tuple_index]
```
Comment (laserkelvin) [inline nvalchemiops/torch/interactions/electrostatics/pme.py]: This mostly mirrors `ewald_summation`, but Ewald uses `_sum_charge_gradients` under `torch.compiler.is_compiling()` while PME does a direct add here. Can we share that helper/pattern, or extract the common result-combiner, so the Ewald and PME slab APIs do not drift on compiled charge-gradient paths?

File: test/interactions/electrostatics/bindings/torch/test_slab.py [hunk 3]
```diff
@@ -867,9 +969,568 @@ def test_hybrid_forward_outputs_match_standard(self, device):
...
+
+        assert positions.dtype == dtype
+        assert charges.dtype == dtype
+        assert cell.dtype == dtype
+        assert energies.dtype == torch.float64
+        assert forces.dtype == dtype
+        assert charge_grads.dtype == torch.float64
+        assert virial.dtype == dtype
+
+    def test_full_pme_slab_output_subsets(self, device):
```
Comment (laserkelvin) [inline test/interactions/electrostatics/bindings/torch/test_slab.py]: Could we extend this output-ordering coverage to include `compute_charge_gradients=True` without forces, plus the mixed three-output cases? That is the fragile part of the new index-based combiner; comparing every flag combination against named component sums would lock down the public PME wrapper behavior.

File: test/interactions/electrostatics/bindings/torch/test_slab.py [hunk 4]
```diff
@@ -867,9 +969,568 @@ def test_hybrid_forward_outputs_match_standard(self, device):
...
+
+    def test_full_pme_slab_requires_pbc(self, device):
+        """PME slab correction requires explicit slab periodicity."""
+        dtype = torch.float64
+
+        positions, charges, cell, _, nl, ptr, shifts = _make_triclinic_ewald_inputs(
+            dtype, device
+        )
+
+        with pytest.raises(ValueError, match="requires an explicit `pbc`"):
```
Comment (laserkelvin) [inline test/interactions/electrostatics/bindings/torch/test_slab.py:1428]: Since `particle_mesh_ewald` is the new public surface, can we also cover the PR #78 `pbc` shape contract through this wrapper: single-system `(3,)` equals `(1, 3)`, and batched `batch_idx` with `(3,)` raises? The helper already has these tests, but wrapper coverage would catch future PME-specific pre-normalization/regression.

File: nvalchemiops/torch/interactions/electrostatics/pme.py [hunk 5]
```diff
@@ -2319,21 +2345,97 @@ def particle_mesh_ewald(
...
+                charges.detach(),
+                cell.detach(),
+                pbc,
+                batch_idx=batch_idx,
+                compute_forces=compute_forces,
+                compute_charge_gradients=True,
+                compute_virial=compute_virial,
+            )
+            slab_raw = slab_out if isinstance(slab_out, tuple) else (slab_out,)
+            slab_index = 0
```
Comment (laserkelvin) [inline nvalchemiops/torch/interactions/electrostatics/pme.py]: Can you add an inline comment to explain what `slab_index` is tracking?

**Review by laserkelvin — CHANGES_REQUESTED** [review]
Minor comments, and then I'll be okay to merge

File: nvalchemiops/torch/interactions/electrostatics/pme.py (hunk 1 above)
Comment (atulcthakur) [reply nvalchemiops/torch/interactions/electrostatics/pme.py:2363]: Failfast is now added in https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/e6ba82d8877b76266f42f6f30ac96c7786b0c030

File: nvalchemiops/torch/interactions/electrostatics/pme.py (hunk 2 above)
Comment (atulcthakur) [reply nvalchemiops/torch/interactions/electrostatics/pme.py]: Same api is now shared across PME and Ewald (https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/e6ba82d8877b76266f42f6f30ac96c7786b0c030)

File: test/interactions/electrostatics/bindings/torch/test_slab.py (hunk 3 above)
Comment (atulcthakur) [reply test/interactions/electrostatics/bindings/torch/test_slab.py]: New tests added in https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/e6ba82d8877b76266f42f6f30ac96c7786b0c030

File: test/interactions/electrostatics/bindings/torch/test_slab.py (hunk 4 above)
Comment (atulcthakur) [reply test/interactions/electrostatics/bindings/torch/test_slab.py:1428]: Done in https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/e6ba82d8877b76266f42f6f30ac96c7786b0c030

File: nvalchemiops/torch/interactions/electrostatics/pme.py (hunk 5 above)
Comment (atulcthakur) [reply nvalchemiops/torch/interactions/electrostatics/pme.py]: Removed `slab_index` and added more explicit unpacking of outputs for better readability https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/e6ba82d8877b76266f42f6f30ac96c7786b0c030

**Comment by atulcthakur** [comment]
/ok to test https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/e6ba82d8877b76266f42f6f30ac96c7786b0c030

**Review by laserkelvin — APPROVED** [review]
LGTM

File: docs/userguide/components/electrostatics.md [hunk 6]
```diff
@@ -488,10 +488,10 @@ Refer to the [Parameter Estimation](parameter-estimation) section for API usage.
 
 #### 2D Slab Correction (PyTorch)
 
-For interfacial systems with two periodic directions and one non-periodic direction,
-PyTorch Ewald can add the Yeh-Berkowitz / Ballenegger slab correction. Pass
-`slab_correction=True` and a boolean `pbc` tensor whose `False` entry marks the
-non-periodic axis:
+For slab-like systems with two periodic directions and one non-periodic direction,
+PyTorch Ewald can add the Yeh-Berkowitz / Ballenegger-Arnold-Cerdà slab
+correction. Pass `slab_correction=True` and a boolean `pbc` tensor whose `False`
```
Comment (zubatyuk) [inline docs/userguide/components/electrostatics.md]: Stress that a `pbc` must have a single `False` item.

File: examples/electrostatics/05_slab_correction_example.py [hunk 7]
```diff
@@ -38,17 +41,22 @@
...
+# The slab correction is available through the high-level Ewald and PME APIs,
+# and as a standalone helper. The standalone helper is useful for debugging,
+# validation, and adding the correction when manually composing components.
 
 from __future__ import annotations
 
 import torch
 
 from nvalchemiops.torch.interactions.electrostatics import (
     apply_slab_correction,
```
Comment (zubatyuk) [inline examples/electrostatics/05_slab_correction_example.py]: Can we change name from `apply_slab_correction` to `slab_correction` or `compute_slab_correction`?

File: examples/electrostatics/05_slab_correction_example.py [hunk 8]
```diff
@@ -218,6 +228,142 @@ def create_cscl_slab_system() -> tuple[
...
+)
+ewald_component_slab_energy, ewald_component_slab_forces = apply_slab_correction(
+    positions=positions,
+    charges=charges,
+    cell=cell,
+    pbc=pbc_slab,
+    compute_forces=True,
+)
+
+manual_ewald_energy = (
```
Comment (zubatyuk) [inline examples/electrostatics/05_slab_correction_example.py]: manual -> total

File: examples/electrostatics/05_slab_correction_example.py [hunk 9]
```diff
@@ -218,6 +228,142 @@ def create_cscl_slab_system() -> tuple[
...
+)
+reciprocal_energy, reciprocal_forces = pme_reciprocal_space(
+    positions=positions,
+    charges=charges,
+    cell=cell,
+    alpha=alpha,
+    mesh_dimensions=mesh_dimensions,
+    compute_forces=True,
+)
+manual_slab_energy, manual_slab_forces = apply_slab_correction(
```
Comment (zubatyuk) [inline examples/electrostatics/05_slab_correction_example.py]: manual_slab_energy, manual_slab_forces -> slab_energy_correction, slab_forces_correction

Comment (atulcthakur) [reply examples/electrostatics/05_slab_correction_example.py]: Fixed

File: examples/electrostatics/05_slab_correction_example.py (hunk 8 above)
Comment (atulcthakur) [reply examples/electrostatics/05_slab_correction_example.py]: Fixed

File: docs/userguide/components/electrostatics.md (hunk 6 above)
Comment (atulcthakur) [reply docs/userguide/components/electrostatics.md]: Fixed

File: examples/electrostatics/05_slab_correction_example.py (hunk 7 above)
Comment (atulcthakur) [reply examples/electrostatics/05_slab_correction_example.py]: This would be a big surgical fix since it'd require changes across docs, api, codes, benchmarks. Will leave unresolved for now and come back to it.

**Comment by atulcthakur** [comment]
/ok to test https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/1346d19b22afaa9d6f94b2f34f0634e00855d74a

**Comment by atulcthakur** [comment]
/ok to test https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/212168687367660fc86c69344e42baddb792f9de

File: examples/electrostatics/05_slab_correction_example.py (hunk 7 above)
Comment (atulcthakur) [reply examples/electrostatics/05_slab_correction_example.py]: Fixed in https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/212168687367660fc86c69344e42baddb792f9de

**Comment by zubatyuk** [comment]
Refactor LGTM

**Comment by atulcthakur** [comment]
/ok to test https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/86/commits/f8a869104d001d9a8f7838150457cd5efebf61ca

**Review by laserkelvin — APPROVED** [review]
